packages/connect-node/src/private/node-universal-client.ts

```diff
@@ -0,0 +1,290 @@
+          request = https.request(req.url, requestOptions);
+        } else {
+          request = http.request(req.url, requestOptions);
+        }
+        request.once("socket", (socket) => {
+          socket.once("connect", connect);
+          socket.once("error", error);
+
+          function connect() {
+            socket.off("error", rejectError);
+            resolve(request);
+          }
+
+          function rejectError(err: Error) {
+            socket.off("connect", connect);
+            error(err);
+          }
+        });
+      }
+    );
+    if (req.body == undefined) {
+      await new Promise<void>((resolve) => {
+        nodeRequest.end(resolve);
+      });
+    } else {
+      await pipeTo(req.body, createRequestSink(nodeRequest), {
+        propagateDownStreamError: true,
+      });
+    }
+    nodeRequest.removeAllListeners();
+    return new Promise<UniversalClientResponse>((resolve, reject) => {
+      nodeRequest.on("error", (args) => {
+        reject(args);
+      });
+      nodeRequest.on("response", (nodeResponse) => {
+        const status = nodeResponse.statusCode ?? 0;
+        const header = nodeHeaderToWebHeader(nodeResponse.headers);
+        const trailer = new Headers();
+        resolve({
+          status,
+          header,
+          body: read(nodeResponse, () => {
+            nodeHeaderToWebHeader(nodeResponse.trailers).forEach(
+              (value, key) => {
+                trailer.set(key, value);
+              }
+            );
+            nodeResponse.removeAllListeners();
+          }),
+          trailer,
+        });
+      });
+    });
+  };
+}
+
+/**
+ * Create an HTTP client using the Node.js `http2` package.
+ *
+ * Note that the client is bound to the authority, and by default, it will close
+ * the HTTP/2 session after the response is read to the end.
+ *
+ * The HTTP client is a simple function conforming to the type UniversalClientFn.
+ * It takes an UniversalClientRequest as an argument, and returns a promise for
+ * an UniversalClientResponse.
+ */
+export function createNodeHttp2Client(
+  authority: URL | string,
+  keepSessionOpen: boolean,
+  http2Options:
+    | http2.ClientSessionOptions
+    | http2.SecureClientSessionOptions
+    | undefined
+): UniversalClientFn {
+  const origin = new URL(authority).origin;
+  let lastSession: http2.ClientHttp2Session | undefined;
+  async function connect(): Promise<http2.ClientHttp2Session> {
+    if (
+      keepSessionOpen &&
+      lastSession !== undefined &&
+      !lastSession.closed &&
+      !lastSession.destroyed
+    ) {
+      return lastSession;
+    }
+    return (lastSession = await new Promise<http2.ClientHttp2Session>(
+      (resolve, reject) => {
+        const s = http2.connect(
+          origin,
+          {
+            ...http2Options,
+          },
+          (s) => resolve(s)
+        );
+        s.on("error", (err) => reject(err));
+      }
+    ));
+  }
```

Comment:
@gilwong00, this ports https://github.com/bufbuild/connect-web/pull/428 - do you spot any obvious issues?

Reply:
same here, looks good!